### PR TITLE
Make AbstractSimpleTransportTestCase more robust

### PR DESCRIPTION
**Pull request by pugnascotia**
`testTransportProfilesWithPortAndHost` assumed that the host only has a single loopback address, which (for reasons I won't go into) isn't necessarily true. Relax this to >=1.

Also convert most of the assertions to use Hamcrest matchers.

### CODE REVIEW

**Review by original-brownbear — DISMISSED** [review]
@pugnascotia could you please split changes into two PRs so that we have a dedicated PR for supporting multiple loopback addresses?

Thanks!

**Comment by pugnascotia** [comment]
I created #45901 with just the loopback change.

**Comment by original-brownbear** [comment]
Thanks @pugnascotia I'm not so sure about moving to Hamcrest, seems like a lot of noise in this codebase (that has a history of instability) just for nicer assertions, but I'll let someone else decide. No strong opinion here :)

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 1]
```diff
@@ -2139,9 +2125,9 @@ public void handleResponse(TransportResponse response) {
             @Override
             public void handleException(TransportException exp) {
                 try {
-                    assertTrue(threadPool.getThreadContext().getResponseHeaders().containsKey("foo.bar"));
-                    assertEquals(1, threadPool.getThreadContext().getResponseHeaders().get("foo.bar").size());
-                    assertEquals("baz", threadPool.getThreadContext().getResponseHeaders().get("foo.bar").get(0));
+                    assertThat(threadPool.getThreadContext().getResponseHeaders(), hasKey("foo.bar"));
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: I don't have a strong opinion about hamcrest for equals checks, but for things like hasKey here, it gives far superior failure information compared to a simple assertTrue/False.

File: test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java [hunk 2]
```diff
@@ -2189,10 +2175,10 @@ public void handleResponse(TransportResponse response) {
             public void handleException(TransportException exp) {
                 try {
                     if (exp instanceof SendRequestTransportException) {
-                        assertTrue(exp.getCause().getClass().toString(), exp.getCause() instanceof NodeNotConnectedException);
+                        assertThat(exp.getCause(), instanceOf(NodeNotConnectedException.class));
                     } else {
                         // here the concurrent disconnect was faster and invoked the listener first
-                        assertTrue(exp.getClass().toString(), exp instanceof NodeDisconnectedException);
+                        assertThat(exp.getClass(), instanceOf(NodeDisconnectedException.class));
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/transport/AbstractSimpleTransportTestCase.java]: This is another case where hamcrest gives much better failure output.

**Review by rjernst — COMMENTED** [review]
I think hamcrest does give better error output for many types of checks. I don't think it matters for assertEquals, but using it consistently is probably good. I believe @jasontedor has been a strong proponent of using hamcrest in the past.

**Comment by pugnascotia** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by pugnascotia** [comment]
I didn't realise that I hadn't had an approval on this when I merged, and @DaveCTurner found an issue, so I'm going to push a revert.

**Pull request merged**